### ISSUE TITLE
Embrace meta data

### ISSUE THREAD

**Issue by mih** [datalad/datalad#618]
We should no longer avoid this issue, hence I am making a proposal.

Let's approach the "whole dataset" level first and only, for now.

I like the following example. It is simple, it is based on standards, it doesn't care much about the format the meta data is stored in.

http://guides.scholarsportal.info/c.php?g=126730&p=829289 (search for "example 1")

Here are the more comprehensive specs from dryad: http://wiki.datadryad.org/images/8/8b/Dryad3.0.pdf

For us the "Dryad Data Package Module" is most relevant, but likely we only need to actively use a subset of that.

Example of such meta data use from datacite: http://data.datacite.org/10.5061/DRYAD.1642B (see "other formats" especially).

Essentially, we only need a mapping from a bunch of terms to some values. Order is irrelevant, additional content is irrelevant (but possible). Storage format is largely arbitrary, as long as we do not invent one (could be CSV, YAML, turtle, anything that allows for a mapping).

Regarding subdataset aggregation: We should stick to what we agreed on a while back -- copy individual files upward into a shadow directory tree, with an optional limit on recursion level. This would be solely for the purpose of obtaining meta data for uninstalled subdatasets. This will keep simple datasets simple, and complex dataset somewhat simple still. Meta data users (including us) can employ this resource and make sense of it through the use of standard terms. Optimized meta data representations for queries (if any) should be compiled on any client machine, but not affect the representation inside a dataset repository.

Thoughts? @yarikoptic @bpoldrack

**Comment by yarikoptic** [comment]
damn -- I thought it was a PR (not just an issue) and got all excited ;)
Interestingly that Dryad Data Package Module  doesn't include "Rights", which seems to be present only at "Dryad Data File Module" level, but we should have rights at least listed.  I will review/compare tomorrow with the datasets descriptors we were planning to use: http://specs.frictionlessdata.io/data-packages/ 
For completeness, here is the whiteboard sketch we had: ![whiteboard](http://www.onerussian.com/tmp/IMG_20160321_110652.jpg)

**Comment by hanke** [comment]
TL;DR: I propose to implement a machinery that doesn't care (much) about semantics, but deals with a flat list of term:value mappings. Domain-specific features and semantics can be build on top of this machinery, later.

Long version:

So wrt this picture, what I am talking about is the datalad-native meta data. I am explicitly not talking about multi-standard, multi-format, mergy-mergy meta data handling. This should be left for later, or we will stay stuck here IMHO. However, whatever we go for now should not become a burden later on.

The reason why my current favorite is a plain (unnested) list of mapping from terms to values is that it is open (can digest any additional stuff that we do not care about (yet)), and that we do not immediately need an ontology mapping from somebodies terms onto ours.

The main downside of this approach is that we cannot go nuts in terms of specificity. It would be difficult to split up meta data on a publication across multiple terms (title, journal, ...) and still be able to specify multiple publications. Or split author names and potential emails. On the other hand this could be considered an advantage, as the maximum possible complexity remains manageable and we could have a finished implementation of something useful before we have invented a complete new meta data standard -- while still being somewhat unconstrained in terms of the actual standards we would support. The lack of "rights" in the data package specs of Dryad is one example, we can simply use it at the dataset level -- the term makes sense nevertheless, and technically it doesn't change things.

Limited complexity would also be achieved by adopting http://specs.frictionlessdata.io/data-packages However, it would come at the cost of being constrained by that specific standard (and possibly of having to implement a parser/validator for that one, I didn't check).

In terms of support for domain-specific standards: I'd say for now we should have something like `--import-from bids` (which points to a BIDS parser plugin) for a single dataset (with the ability to store this choice in some `meta.cfg`), but right now we should not attempt to parse, filter, merge, or aggregate any of that for a superdataset. Aggregation should be implemented as planned by copying our own metadata (which may have had its origin in BIDS) upward -- for now.

The meta data update procedure is a little tricky, though (or I just don't see it clearly yet). Suppose a dataset has many contributors, like this one: http://datadryad.org/resource/doi:10.5061/dryad.77471?show=full and now I want to update a name, I'd need something like:

````
--replace TERM VALUE_OLD VALUE_NEW
````

In order to be able to find and change what I want. While for an author this would still be manageable to do by hand in the cmdline, it becomes impractical for, e.g. an abstract. This boils down to the question of whether it is worth offering a programmatic interface to modification (other than a text editor) at all?

Things could be simplified, though. There is likely just one abstract, while there are often many authors. Moreover, as we would have a plain list of mappings, there is the option to simply address them by a global integer index (or a local, per-term, index, e.g. the 5th author). Some syntactical sugar could get us a long way, IMHO.

````
--replace dc.contributor.author#5 "my name"

--replace #16 "i don't even know what I am"

--remove dc.description

--remove dc.contributor.author Michael
````

**Comment by hanke** [comment]
About to push API proposal in a PR.

**Comment by yarikoptic** [comment]
Fwiw staying with the editor is OK with me, we just need some API for  initial initialization of basic meta by the user
As for format - I am ok with anything we could reliably and quickly read without demanding additional heavy library

**Comment by hanke** [comment]
@yarikoptic Please have a look at #624

**Comment by hanke** [comment]
Nothing is set about a format yet, but we would be dealing with plain key value pairs -- a very simple format would suffice -- right now I don't see the need for any library.

**Comment by bpoldrack** [comment]
Generally I agree on having a "machinery, that doesn't care much about semantics". I also agree on the kind of propagation @hanke proposed.

The most important thing to me are:
- datalad-native meta data needs to (technically) be agnostic  to underlying meta data formats
- propagation should be configurable
- the logical structure of datalad native meta data needs to be very flexible, since we don't know yet, what the things are, that we want to support sooner or later

My concern is the last point. I'm not sure, whether we achieve this by a flat list of terms. May be I just didn't entirely get, what the idea is. How is defining a list of terms different from defining our own ontology? Furthermore, if it is not supposed to be "nested" we need to provide several "tables" (or sections in the file), that can be linked, in order to allow for more structured meta data later on, I guess. If the keys of those key, value pairs are too simple, it will be difficult to introduce structured things later on.

Either way, most important is `datalad-metadata-format-version` marker, allowing for conversions later on. ;)

**Comment by hanke** [comment]
@bpoldrack I don't think we need a format marker (little exaggeration to drive my point home).

We are not defining our own terms or using any ontology, we have pairs of text items, invented by users -- for the crawler we could define one, or adopt Dryad's, but we do not need to impose anything on a user.

We don't want a tree, because it is hard, and sometimes impossible to merge them -- difficult enough with a list.

I believe that if we try to come up with a system that could store anything, we will fail to get something useful up and running in a reasonable time. It is easy to come up with a use case that cannot be supported by my proposal, but we need to look whether such use case is relevant enough to justify a more complicated system.

My proposal is about metadata for a "dataset as a whole" -- that alone makes most desired metadata descriptions impossible.

My approach was to go find a metadata scheme that is used in the wild and see whether we can adopt it to move forward. I believe the "some terms form DC in a flat list" used by Dryad is a good candidate.

I no longer see the need for a system that is extensible enough to capture anything. If datasets come with metadata already, we should simply _store_ it as is. RIght now I think that up-propagation of anything other than top-level dataset descriptions is not desirable, hence no need to manage anything but that.

**Comment by bpoldrack** [comment]
To be more clear about "structured" meta data:
A simple key, value pair like `dc:contributor.author=Michael Hanke` doesn't tell what entity it is about. For now we would naturally assume, that this declares one author of the dataset. In terms of triples: The SUBJECT is implicitly the dataset itself (may be derived from the location, the file is stored at). If we are to extend metadata at some point, we might want to include subentities, that are not subdatasets as "subjects" of a term. A file for example or a set of files, that have different authorship or different license. Semantically this would still be the same term of the DC ontology. But we need to allow for different subjects. Which leads to the need to identify them.

**Comment by bpoldrack** [comment]
Didn't see your previous post before sending mine.

Well, if we restrict this approach to "dataset as a whole" my point is kind of invalid. I agree. We just should make sure, we are able to add a more complex layer without shooting ourselves, I guess.

**Comment by yarikoptic** [comment]
Here is some kind of a spec from biocaddie https://docs.google.com/document/d/1hVcYRleE6-dFfn7qbF9Bv1Ohs1kTF6a8OwWUvoZlDto/edit  which also links to other existing schemas

**Comment by hanke** [comment]
Thought about this some more. Came down to these points:

1. We should not perform meta data mangling/copying for non-super datasets. Instead, we should just identify the type of meta data (format) available in a dataset (via a config file). If these meta data are to be read (e.g. to put them into a query-optimized form) the matching parser should read them from their raw form. We should then have a parser for BIDS, and some others on a by-demand basis. Bottom line: aggregate data is automatically generated and not meant to be edited, or duplicated within one dataset.
2. When we aggregate meta data in super datasets we want to have maximum flexibility to be able to store anything, without causing complications via "hierarchical" formats (YAML, JSON, or anything that isn't a flat list of statements), or from being forced to have one file per subdataset, etc. At the same time, we want a format that has a simple structure to avoid having to implement merge-helpers and such -- ideally a line-based format, so even things like a quick grep yield meaningful results.

I found the N-triples format to offer the best compromise.

https://en.wikipedia.org/wiki/N-Triples

Full spec: https://www.w3.org/TR/n-triples

This gives us the expressiveness of a graph, but in a line-based format. Queries can be as simple as a `grep`. Merge-friendliness seems to be there.

We could explore a possible naming convention for "blank nodes" to add meaning to the statement's subject components without having to resort to full URLs.

Looking at the spec it seems rather simple to read and write such files without an RDF library, or even being aware of the graph-nature of the data. Conversion into a tabular for seems simple too.

If we were to explore graph queries in the future, RDFlib has built-in support for this format. But again, no need for RDFlib for basic operation (incl. modification).

**Comment by yarikoptic** [comment]
2. "When we aggregate meta data ...  ideally a line-based format, so even things like a quick grep yield meaningful results"  
I think that when we aggregate we should just leave those formats in original form, and just cache them in the super-dataset(s).  This would allow to avoid making meta-data "datalad-version-specific" since any aggregation might then be dependent on datalad version.  It is only for querying operations, we should generate some local representation of available meta-data, as available within super-datasets which aggregate data from sub-datasets.

The question is though which format to choose for datasets meta-data  we provide and/or overlay on top of existing meta-data.   Before I start whining about n-triples, I would appreciate an example for some dataset(s) to get a better sense.  Ideally, I would like to see some human-readable and editable format without much of boiler plate... ideally altogether reusing terms or the entire specification which already exists, so we don't brew a new beast.

**Comment by hanke** [comment]
@yarikoptic I started crafting examples and code.

re aggregation in native form: This is tricky. Depending on the source format, it may be impossible to just copy things over without some kind of extraction/compacting.

re native format: I am still leaning heavily towards a subset of the GNIT-style meta data files:  https://en.wikipedia.org/wiki/README

Starting crafting a parser already.

**Comment by yarikoptic** [comment]
On Mon, 01 Aug 2016, Michael Hanke wrote:

>    @yarikoptic I started crafting examples and code.

>    re aggregation in native form: This is tricky. Depending on the source
>    format, it may be impossible to just copy things over without some kind of
>    extraction/compacting.

>    re native format: I am still leaning heavily towards a subset of the
>    GNIT-style meta data files: https://en.wikipedia.org/wiki/README

FWIW

1. Among relevant I only see README (description), COPYING/LICENSE (free
form full license text? how to aggregate for concise
representation), AUTHORS/THANKS

2. They are also lack any "standard" place for common to our usecase
placeholders for citations / dois, possible keywords, ...

3. those aren't machine readable per se since lacking clear standard
on formats of each one of those.  So would be hard to verify correct
formatting, present in the "dataset view", etc

4. Spreading meta-data among files would not be of any help IMHO, it
might hurt during aggregations.

NB I sometimes like how rpm source packages single file for simple
packages feels more adequate than a collection of >5 files under debian/
... The same here -- not sure if aiming to catch a common use case of
having README+LICENSE/COPYING wouldn't hurt us in the longer run

>    Starting crafting a parser already.

Sure thing we can provide some 'import' facility from GNU-style
specifications, but I wouldn't aim for it to be the primary way to
provide meta-data for the datasets.

Related -- many upstream datasets already provide some README and/or
COPYING, but they might be not easily "aggregatable" (too long/too
short/...).  Should we then patch them?

-- 
Yaroslav O. Halchenko
Center for Open Neuroscience     http://centerforopenneuroscience.org
Dartmouth College, 419 Moore Hall, Hinman Box 6207, Hanover, NH 03755
Phone: +1 (603) 646-9834                       Fax: +1 (603) 646-1419
WWW:   http://www.linkedin.com/in/yarik

**Comment by hanke** [comment]
To clarify:

I would not aim for a single "standard" place for meta data, but rather support reading from various formats. Any such parser would be able to read a certain set of metadata items. I proposed a set here:

https://github.com/hanke/datalad/blob/nf-metaparser/datalad/metadata/common.py

Regardless how widespread metadata is in the raw dataset, it would be converted into a flat list of n-triples upon request (e.g. for aggregation, or for query). When aggregating, multiple of such lists can be concatenated and stored in a single, or multiple files (whatever suits our needs).

Hence when the crawler creates datasets, it can chose to write metadata in any format for which we have parsers. For openfmri datasets it may be possible to not write additional metadata at all. For other sources where metadata needs to be scraped from outside the datasets themselves, the most appropriate format can be chosen.

Here is a PR that is working towards such a system: #682

We could still think of creating a standardized copy of the relevant metadata even in non super-datasets, in case annexed files would have to be retrieved to be able to access the metadata. But this could be a per-dataset (or per-portal) decision.

**Comment by hanke** [comment]
Oh and to be clear: We can have any additional metadata format supported to just be able to enter our minimally supported datalad-specific metadata item set. CSV, JSON, whatever standard, all fine.

**Comment by mih** [comment]
Something along those lines is implemented in #682
